doc/source/serialization.rst

```diff
@@ -64,6 +47,107 @@ Serialization notes
 
 - Lock objects are mostly unserializable, because copying a lock is meaningless and could cause serious concurrency problems. You may have to come up with a workaround if your object contains a lock.
 
+Custom Serialization
+____________________
+
+Sometimes you may want to customize your serialization process because
+the default serializer used by Ray (pickle5 + cloudpickle) does
+not work for you (fail to serialize some objects, too slow for certain objects, etc.).
+
+There are at least 3 ways to define your custom serialization process:
+
+1. If you want to customize the serialization of a type of objects,
+   and you have access to the code, you can define ``__reduce__``
+   function inside the corresponding class. This is commonly done
+   by most Python libraries. Example code:
+
+.. code-block:: python
+
+  import ray
+  import sqlite3
+
+  ray.init()
+
+  class DBConnection:
+      def __init__(self, path):
+          self.path = path
+          self.conn = sqlite3.connect(path)
+
+      # without '__reduce__', the instance is unserializable.
+      def __reduce__(self):
+          deserializer = DBConnection
+          serialized_data = (self.path,)
+          return deserializer, serialized_data
+
+  original = DBConnection("/tmp/db")
+  print(original.conn)
+
+  copied = ray.get(ray.put(original))
+  print(copied.conn)
+
+2. If you want to customize the serialization of a type of objects,
+   but you cannot access or modify the corresponding class, you can
+   register the class with the reducer/serializer you use:
+
+.. code-block:: python
+
+  import ray.serialization
+  import threading
+
+  class A:
+      def __init__(self, x):
+          self.x = x
+          self.lock = threading.Lock()  # could not serialize!
+
+  ray.get(ray.put(A(1)))  # fail!
+
+  def custom_serializer(a):
+      return a.x
+
+  def custom_deserializer(b):
+      return A(b)
+
+  # Here we show two different APIs for customized serialization.
+
+  # API 1: serializer & deserializer
+  ray.serialization.register_serializer(A, custom_serializer, custom_deserializer)
+  ray.get(ray.put(A(1)))  # success!
+
+  # API 2: reducer
+  # A reducer is serializer and deserializer combined. Sometimes it is more convenient.
+  def custom_reducer(a):
+      return A, (a.x,)  # equivalent to 'custom_deserializer, (custom_serializer(a),)'
+
+  ray.serialization.register_reducer(A, custom_reducer)
+  ray.get(ray.put(A(1)))  # success!
+
+3. If you want to customize the serialization of a specific object,
```

Comment:
Do we really need this? Seems like potentially overkill.

Reply:
It should be useful as the final fallback if we cannot serialized something according to the type, because the object type is hacked. Most common cases are types derived from metaclasses (for example, a lot of dataclass types provided by pydantic). This API ensures that the serialization can success in any condition.

Reply:
I agree, let's remove this from the documentation and also remove the `wrap_reducer` method from below :)

Reply:
(users could always define this type themselves if they **really** need it)